### PR TITLE
1388 xml serializer assembly load context awareness

### LINKED ISSUE #1388: AssemblyLoadContext crash when collectible assembly use XmlSerializer
[EDIT]
Investigation of this issue is complete and has resulted in 5 additional issues being opened.
[See this comment below](https://github.com/dotnet/corefx/issues/41286#issuecomment-562741471) for links to each of those issues.

We will keep this issue open and update it as those other issues are fixed.
Please add any additional comments to those issues.

**********************************************************************************************************
When creating a collectible AssemblyLoadContext and then use `System.Xml.Serialization.XmlSerializer` inside of a collectible assembly you will get a crash.

I have a repro project: https://github.com/uffebjorklund/AssemblyLoadContextBug

Have tried a few different things
 - If creating the ALC as `collectible = false` there is no crash
 - If the `XmlSerializer` code is removed from the collectible assembly there is no crash
 - The crash will occur regardless of using `netstandard2.0` or `netstandard2.1` on the collectible assembly

## Exception
`A non-collectible assembly may not reference a collectible assembly`

## Stacktrace
```
at System.Reflection.Emit.ModuleBuilder.GetTypeRef(QCallModule module, String strFullName, QCallModule refedModule, String strRefedModuleFileName, Int32 tkResolution)
at System.Reflection.Emit.ModuleBuilder.GetTypeRefNested(Type type, Module refedModule, String strRefedModuleFileName)
at System.Reflection.Emit.ModuleBuilder.GetTypeTokenWorkerNoLock(Type type, Boolean getGenericDefinition)
at System.Reflection.Emit.ModuleBuilder.GetTypeTokenInternal(Type type, Boolean getGenericDefinition)
at System.Reflection.Emit.ILGenerator.Emit(OpCode opcode, Type cls)
at System.Xml.Serialization.CodeGenerator.Castclass(Type target)
at System.Xml.Serialization.CodeGenerator.InternalConvert(Type source, Type target, Boolean isAddress)
at System.Xml.Serialization.SourceInfo.InternalLoad(Type elementType, Boolean asAddress)
at System.Xml.Serialization.XmlSerializationWriterILGen.WriteElement(SourceInfo source, ElementAccessor element, String arrayName, Boolean writeAccessor)
at System.Xml.Serialization.XmlSerializationWriterILGen.WriteElements(SourceInfo source, String enumSource, ElementAccessor[] elements, TextAccessor text, ChoiceIdentifierAccessor choice, String arrayName, Boolean writeAccessors, Boolean isNullable)
at System.Xml.Serialization.XmlSerializationWriterILGen.WriteMember(SourceInfo source, String choiceSource, ElementAccessor[] elements, TextAccessor text, ChoiceIdentifierAccessor choice, TypeDesc memberTypeDesc, Boolean writeAccessors)
at System.Xml.Serialization.XmlSerializationWriterILGen.GenerateTypeElement(XmlTypeMapping xmlTypeMapping)
at System.Xml.Serialization.XmlSerializationWriterILGen.GenerateElement(XmlMapping xmlMapping)
at System.Xml.Serialization.TempAssembly.GenerateRefEmitAssembly(XmlMapping[] xmlMappings, Type[] types, String defaultNamespace)
at System.Xml.Serialization.TempAssembly..ctor(XmlMapping[] xmlMappings, Type[] types, String defaultNamespace, String location)
at System.Xml.Serialization.XmlSerializer.GenerateTempAssembly(XmlMapping xmlMapping, Type type, String defaultNamespace, String location)
at System.Xml.Serialization.XmlSerializer..ctor(Type type, String defaultNamespace)
```

## dotnet --info
```
.NET Core SDK (reflecting any global.json):
 Version:   3.0.100
 Commit:    04339c3a26

Runtime Environment:
 OS Name:     Mac OS X
 OS Version:  10.13
 OS Platform: Darwin
 RID:         osx.10.13-x64
 Base Path:   /usr/local/share/dotnet/sdk/3.0.100/

Host (useful for support):
  Version: 3.0.0
  Commit:  7d57652f33

.NET Core SDKs installed:
  2.1.301 [/usr/local/share/dotnet/sdk]
  2.1.302 [/usr/local/share/dotnet/sdk]
  2.1.400 [/usr/local/share/dotnet/sdk]
  2.1.401 [/usr/local/share/dotnet/sdk]
  2.1.403 [/usr/local/share/dotnet/sdk]
  2.1.500 [/usr/local/share/dotnet/sdk]
  2.2.100-preview3-009430 [/usr/local/share/dotnet/sdk]
  2.2.101 [/usr/local/share/dotnet/sdk]
  2.2.102 [/usr/local/share/dotnet/sdk]
  2.2.103 [/usr/local/share/dotnet/sdk]
  2.2.105 [/usr/local/share/dotnet/sdk]
  2.2.106 [/usr/local/share/dotnet/sdk]
  2.2.203 [/usr/local/share/dotnet/sdk]
  2.2.300 [/usr/local/share/dotnet/sdk]
  3.0.100-alpha1-009708 [/usr/local/share/dotnet/sdk]
  3.0.100-preview-010184 [/usr/local/share/dotnet/sdk]
  3.0.100-preview3-010431 [/usr/local/share/dotnet/sdk]
  3.0.100-preview8-013656 [/usr/local/share/dotnet/sdk]
  3.0.100-preview9-014004 [/usr/local/share/dotnet/sdk]
  3.0.100 [/usr/local/share/dotnet/sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.1 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.2 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.3 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.5 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.6 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.0-preview3-35497 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.0 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.1 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.3 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.4 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.5 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 3.0.0-alpha1-10062 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.1 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.2 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.3 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.5 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.6 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.0-preview3-35497 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.0 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.1 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.3 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.4 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.5 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.0.0-alpha1-10062 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.0.0-preview-19075-0444 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.0.0-preview3-19153-02 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.0.0-preview8.19405.7 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.0.0-preview9.19424.4 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.0.0 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.1 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.2 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.3 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.5 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.6 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.0-preview3-27014-02 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.0 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.1 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.3 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.4 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.5 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.0.0-preview-27324-5 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.0.0-preview1-27029-03 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.0.0-preview3-27503-5 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.0.0-preview9-19423-09 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.0.0 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
```

### PR DESCRIPTION
**Pull request by StephenMolloy**
Makes XmlSerializer aware of AssemblyLoadContexts

When serializing types that are in a different ALC, the serializer
should also operate in that same context. And not keep any
hard references that would prevent the context from being
unloaded.

Fixes #1388, #598, #599

### CODE REVIEW

File: src/libraries/System.Private.Xml/src/System/Xml/Serialization/Compilation.cs [hunk 1]
```diff
@@ -710,8 +711,12 @@ internal void Add(string? ns, object o, TempAssembly assembly)
...
                 if (_cache.TryGetValue(key, out tempAssembly) && tempAssembly == assembly)
                     return;
-                Dictionary<TempAssemblyCacheKey, TempAssembly> _copy = new Dictionary<TempAssemblyCacheKey, TempAssembly>(_cache); // clone
-                _copy[key] = assembly;
+                ConditionalWeakTable<TempAssemblyCacheKey, TempAssembly> _copy = new ConditionalWeakTable<TempAssemblyCacheKey, TempAssembly>(); // clone
+                foreach (KeyValuePair<TempAssemblyCacheKey, TempAssembly> kvp in _cache)
+                {
+                    _copy.Add(kvp.Key, kvp.Value);
+                }
+                _copy.Add(key, assembly);
```
Comment (stephentoub) [inline src/libraries/System.Private.Xml/src/System/Xml/Serialization/Compilation.cs:719]: The `key` instance being constructed here isn't rooted by anything else.  Doesn't this change mean that things added to the cache will end up being collected the next time the GC runs?  Is that what's desired?

Comment (mconnew) [reply src/libraries/System.Private.Xml/src/System/Xml/Serialization/Compilation.cs:719]: @stephentoub you are correct. The ConditionalWeakTable needs to use the Type object as the key, and the value probably needs to be a `Dictionary<TempAssemblyCacheKey, TempAssembly>()`. So we first lookup by Type, and then in the returned dictionary, do the lookup by TempAssemblyCacheKey. We could use a list, but then we introduce the edge case of if someone has code to create a serializer with arbitrary namespaces it could slow down the lookup.  
If this goes to servicing instead of initial release, I think it might be worth having two caches, one for types where the ALC is not collectible, and one for types where the ALC is collectible. ALC has an IsCollectible property. I don't know what the overhead is for ConditionalWeakTable, but I suspect it's quite a lot higher than it is for Dictionary.

File: src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlSerializationWriter.cs [hunk 2]
```diff
@@ -1465,13 +1466,13 @@ internal static class DynamicAssemblies
     {
         private static readonly Hashtable s_nameToAssemblyMap = new Hashtable();
         private static readonly Hashtable s_assemblyToNameMap = new Hashtable();
-        private static readonly Hashtable s_tableIsTypeDynamic = Hashtable.Synchronized(new Hashtable());
+        private static readonly ConditionalWeakTable<Type, object> s_tableIsTypeDynamic = new ConditionalWeakTable<Type, object>();
```
Comment (mconnew) [inline src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlSerializationWriter.cs:1470]: The previous Hashtable was wrapped in Hashtable.Synchronized, but the ConditionalWeakTable isn't. Is it thread safe or do you need to make all access be wrapped in a lock?

File: src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlSerializer.cs [hunk 3]
```diff
@@ -391,7 +393,12 @@ public void Serialize(XmlWriter xmlWriter, object? o, XmlSerializerNamespaces? n
                     }
                 }
                 else
-                    _tempAssembly.InvokeWriter(_mapping, xmlWriter, o, namespaces == null || namespaces.Count == 0 ? DefaultNamespaces : namespaces, encodingStyle, id);
+                {
+                    using (AssemblyLoadContext.EnterContextualReflection(_tempAssembly._assembly))
+                    {
```
Comment (mconnew) [inline src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlSerializer.cs:398]: Why was this needed here? I wouldn't expect any code generation on this call path as the assembly has already been loaded/created.

**Comment by mconnew** [comment]
>         {

Each of the XmlMapping instances could be for a type in a different ALC. I think this is still going to break as it iterates through the types for each of the mappings and generates the serializer for them. If they are in a different ALC then type (which might be null), then it will throw. To fix this looks like it would require quite a bit of refactoring. Instead it would be worth detecting this situation and throwing an exception saying this isn't supported and all XmlMappings need to be for the same ALC.

---
Refers to: src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlSerializer.cs:561 in 8146852. [](commit_id = 8146852e3ae0b9051720359291700d1e6e51a662, deletion_comment = False)

File: src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlSerializer.cs [hunk 4]
```diff
@@ -683,7 +696,7 @@ private static XmlSerializer[] GetSerializersFromCache(XmlMapping[] mappings, Ty
                 if (!s_xmlSerializerTable.TryGetValue(type, out typedMappingTable))
                 {
                     typedMappingTable = new Dictionary<XmlSerializerMappingKey, XmlSerializer>();
-                    s_xmlSerializerTable[type] = typedMappingTable;
+                    s_xmlSerializerTable.Add(type, typedMappingTable);
                 }
```
Comment (mconnew) [inline src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlSerializer.cs:700]: Similar comment to elsewhere, should we have 2 backing stores, one for collectible and one for non-collectible? I don't know what the performance will be like on the TryGetValue.

File: src/libraries/System.Private.Xml/tests/XmlSerializer/XmlSerializerTests.cs [hunk 5]
```diff
@@ -1960,6 +1960,42 @@ public static void Xml_TypeWithSpecialCharacterInStringMember()
...
+        for (int i = 0; weakRef.IsAlive && i < 10; i++)
+        {
+            GC.Collect();
+            GC.WaitForPendingFinalizers();
+        }
+        Assert.True(!weakRef.IsAlive);
+    }
+
+        private static void ExecuteAndUnload(string assemblyname, string typename, out WeakReference wref)
+    {
```
Comment (mconnew) [inline src/libraries/System.Private.Xml/tests/XmlSerializer/XmlSerializerTests.cs:1983]: Nit: Indentation

File: src/libraries/System.Private.Xml/tests/XmlSerializer/XmlSerializerTests.cs [hunk 6]
```diff
@@ -1960,6 +1960,42 @@ public static void Xml_TypeWithSpecialCharacterInStringMember()
...
     }
 
+    [Fact]
+    public static void Xml_TypeInCollectibleALC()
+    {
+        WeakReference weakRef;
+
+#if !XMLSERIALIZERGENERATORTESTS
+        ExecuteAndUnload("System.Collections.Specialized", "System.Collections.Specialized.StringCollection", out weakRef);
+#else
```
Comment (mconnew) [inline src/libraries/System.Private.Xml/tests/XmlSerializer/XmlSerializerTests.cs:1970]: Are you sure this is getting loaded into the collectible ALC? I remember something about types from .NET itself will get loaded into the default ALC regardless unless you provide a custom assembly resolver. Can you verify that the loaded assembly is in your collectible ALC?

File: src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlSerializationWriter.cs (hunk 2 above)
Comment (StephenMolloy) [reply src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlSerializationWriter.cs:1470]: Yes, it is thread safe. https://docs.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.conditionalweaktable-2?view=net-5.0#thread-safety

File: src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlSerializer.cs (hunk 3 above)
Comment (StephenMolloy) [reply src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlSerializer.cs:398]: We can discuss in our weekly meeting, but if we're executing a read/write on a serializer that is loaded in a specific context, wouldn't we want any by/products of that execution to also live within that context?

**Review by mconnew — APPROVED** [review]
:shipit:

**Review by HongGit — APPROVED** [review]
